File: packages/typescript/src/validator.ts
import { Parser, type AST } from 'node-sql-parser';
import { CostEstimate, CostEstimator, CostLevel } from './cost';
import { LimitEnforcer, isLimitOk } from './limits';
import { ValidationResult } from './result';
import { SecurityConfig, type SecurityConfigOptions } from './security';
import { runSecurityRules } from './rules';

/**
 * Validation mode.
 */
export type Mode = 'read_only' | 'write_safe' | 'custom';

/**
 * Options for creating a Validator.
 */
export interface ValidatorOptions {
  /** Validation mode (default: "read_only") */
  mode?: Mode;
  /** Optional list of table names that can be accessed */
  allowedTables?: string[];
  /** For custom mode: statements to allow */
  allowedStatements?: string[];
  /** For custom mode: statements to block */
  blockedStatements?: string[];
  /** SQL dialect for parsing */
  dialect?: string;
  /** Security rule configuration (true/false/SecurityConfig/SecurityConfigOptions) */
  securityConfig?: boolean | SecurityConfig | SecurityConfigOptions;
  /** Maximum allowed LIMIT value */
  maxRows?: number;
  /** If true, SELECT queries without LIMIT are rejected */
  requireLimit?: boolean;
  /** If true, estimate query cost/complexity */
  estimateCost?: boolean;
  /** If true, block queries with HIGH or EXTREME cost */
  blockHighCost?: boolean;
  /** Maximum allowed cost level */
  maxCostLevel?: CostLevel | keyof typeof CostLevel;
}

// Statement types allowed in each mode
const READ_ONLY_ALLOWED = new Set(['SELECT']);
const WRITE_SAFE_ALLOWED = new Set(['SELECT', 'INSERT', 'UPDATE']);
const WRITE_SAFE_BLOCKED = new Set([
  'DELETE',
  'DROP',
  'TRUNCATE',
  'ALTER',
  'CREATE',
  'GRANT',
  'REVOKE',
]);

/**
 * SQL query validator.
 *
 * For repeated validations with the same configuration, create a Validator
 * instance rather than calling validate() each time.
 *
 * @example
 * ```typescript
 * const validator = new Validator({
 *   mode: "read_only",
 *   allowedTables: ["products", "categories"]
 * });
 *
 * validator.validate("SELECT * FROM products").isSafe  // true
 * validator.validate("SELECT * FROM users").isSafe     // false
 * ```
 */
export class Validator {
  private readonly mode: Mode;
  private readonly allowedTables: Set<string> | null;
  private readonly allowedStatements: Set<string> | null;
  private readonly blockedStatements: Set<string> | null;
  private readonly dialect: string;
  private readonly securityConfig: SecurityConfig | null;
  private readonly parser: Parser;
  private readonly limitEnforcer: LimitEnforcer | null;
  private readonly costEstimator: CostEstimator | null;
  private readonly maxCostLevel: CostLevel | null;

  constructor(options: ValidatorOptions = {}) {
    this.mode = options.mode ?? 'read_only';
    this.dialect = options.dialect ?? 'postgresql';
    this.parser = new Parser();

    // Handle allowed tables (case-insensitive)
    if (options.allowedTables) {
      this.allowedTables = new Set(options.allowedTables.map((t) => t.toLowerCase()));
    } else {
      this.allowedTables = null;
    }

    // Handle security config
    if (options.securityConfig === false) {
      this.securityConfig = null;
    } else if (options.securityConfig === true || options.securityConfig === undefined) {
      this.securityConfig = new SecurityConfig();
    } else if (options.securityConfig instanceof SecurityConfig) {
      this.securityConfig = options.securityConfig;
    } else {
      // Plain object passed - convert to SecurityConfig
      this.securityConfig = new SecurityConfig(options.securityConfig);
    }

    // Initialize statement rules based on mode
    this.allowedStatements = this.initAllowedStatements(options);
    this.blockedStatements = this.initBlockedStatements(options);

    // Set up limit enforcer
    if (options.maxRows !== undefined || options.requireLimit) {
      this.limitEnforcer = new LimitEnforcer({
        maxRows: options.maxRows,
        requireLimit: options.requireLimit,
      });
    } else {
      this.limitEnforcer = null;
    }

    // Set up cost estimator
    const shouldEstimateCost =
      options.estimateCost || options.blockHighCost || options.maxCostLevel !== undefined;
    this.costEstimator = shouldEstimateCost ? new CostEstimator() : null;

    // Set up max cost level
    if (options.maxCostLevel !== undefined) {
      this.maxCostLevel =
        typeof options.maxCostLevel === 'string'
          ? CostLevel[options.maxCostLevel]
          : options.maxCostLevel;
    } else if (options.blockHighCost) {
      this.maxCostLevel = CostLevel.MEDIUM;
    } else {
      this.maxCostLevel = null;
    }
  }

  private initAllowedStatements(options: ValidatorOptions): Set<string> | null {
    switch (this.mode) {
      case 'read_only':
        return READ_ONLY_ALLOWED;
      case 'write_safe':
        return WRITE_SAFE_ALLOWED;
      case 'custom':
        if (options.allowedStatements) {
          return new Set(options.allowedStatements.map((s) => s.toUpperCase()));
        }
        return null;
      default:
        return READ_ONLY_ALLOWED;
    }
  }

  private initBlockedStatements(options: ValidatorOptions): Set<string> | null {
    switch (this.mode) {
      case 'read_only':
        return null; // read_only uses allowlist only
      case 'write_safe':
        return WRITE_SAFE_BLOCKED;
      case 'custom':
        if (options.blockedStatements) {
          return new Set(options.blockedStatements.map((s) => s.toUpperCase()));
        }
        return null;
      default:
        return null;
    }
  }

  /**
   * Validate a SQL query.
   */
  validate(sql: string): ValidationResult {
    // Check for empty/whitespace-only queries
    if (!sql || !sql.trim()) {
      return ValidationResult.unsafe('Query is empty or contains only whitespace');
    }

    try {
      // Parse the SQL
      const ast = this.parser.astify(sql, { database: this.mapDialect(this.dialect) });
      const statements = Array.isArray(ast) ? ast : [ast];

      if (statements.length === 0) {
        return ValidationResult.unsafe('No valid SQL statements found');
      }

      const warnings: string[] = [];
      let combinedCost: CostEstimate | undefined;
      let limitValue: number | undefined;

      // Validate each statement
      for (const stmt of statements) {
        // 1. Check statement type
        const typeResult = this.checkStatementType(stmt);
        if (!typeResult.isSafe) return typeResult;

        // 2. Check table allowlist
        const tableResult = this.checkTables(stmt, sql);
        if (!tableResult.isSafe) return tableResult;

        // 3. Check limit enforcement
        if (this.limitEnforcer) {
          const limitResult = this.limitEnforcer.check(stmt);
          if (!isLimitOk(limitResult)) {
            return ValidationResult.unsafe(limitResult.reason!, {
              statementType: this.getStatementType(stmt),
              tables: this.extractTables(stmt),
            });
          }
          if (limitResult.limitValue !== undefined) {
            limitValue = limitResult.limitValue;
          }
        }

        // 4. Check cost estimation
        if (this.costEstimator) {
          const cost = this.costEstimator.estimate(stmt);

          // Track highest cost across statements
          if (!combinedCost || cost.level > combinedCost.level) {
            combinedCost = cost;
          }

          // Check if cost exceeds maximum
          if (this.maxCostLevel !== null && cost.level > this.maxCostLevel) {
            return ValidationResult.unsafe(
              `Query cost (${CostLevel[cost.level]}) exceeds maximum allowed (${CostLevel[this.maxCostLevel]})`,
              {
                statementType: this.getStatementType(stmt),
                tables: this.extractTables(stmt),
                cost,
              }
            );
          }

          // Add warnings for high-cost queries that aren't blocked
          if (cost.level >= CostLevel.HIGH && this.maxCostLevel === null) {
            warnings.push(
              `High cost query (${CostLevel[cost.level]}): ${cost.factors.join(', ')}`
            );
          }
        }
      }

      // 5. Run security rules (on entire query)
      if (this.securityConfig) {
        const securityResult = runSecurityRules(sql, ast, this.securityConfig);
        if (!securityResult.isSafe) return securityResult;
      }

      // All checks passed
      const firstStmt = statements[0] ?? null;
      return ValidationResult.safe({
        statementType: this.getStatementType(firstStmt),
        tables: this.extractAllTables(statements),
        cost: combinedCost,
        limitValue,
        warnings,
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return ValidationResult.unsafe(`SQL parse error: ${message}`);
    }
  }

  private mapDialect(dialect: string): string {
    // Map common dialect names to node-sql-parser database names
    const dialectMap: Record<string, string> = {
      postgres: 'postgresql',
      postgresql: 'postgresql',
      mysql: 'mysql',
      sqlite: 'sqlite',
      snowflake: 'snowflake',
      bigquery: 'bigquery',
      redshift: 'redshift',
      hive: 'hive',
      spark: 'spark',
      trino: 'trino',
      presto: 'trino',
      flinksql: 'flinksql',
      transactsql: 'transactsql',
      mssql: 'transactsql',
      sqlserver: 'transactsql',
    };
    return dialectMap[dialect.toLowerCase()] ?? 'postgresql';
  }

  private getStatementType(stmt: AST | null): string {
    if (!stmt) return 'UNKNOWN';
    // node-sql-parser uses 'type' for the statement type
    return (stmt as any).type?.toUpperCase() ?? 'UNKNOWN';
  }

  private checkStatementType(stmt: AST): ValidationResult {
    const stmtType = this.getStatementType(stmt);

    // Check blocklist first (blocked takes precedence)
    if (this.blockedStatements?.has(stmtType)) {
      return ValidationResult.unsafe(
        `Statement type '${stmtType}' is blocked in ${this.mode} mode`
      );
    }

    // Check allowlist
    if (this.allowedStatements && !this.allowedStatements.has(stmtType)) {
      return ValidationResult.unsafe(
        `Statement type '${stmtType}' is not allowed in ${this.mode} mode`
      );
    }

    return ValidationResult.safe();
  }

  private checkTables(stmt: AST, _sql: string): ValidationResult {
    if (!this.allowedTables) return ValidationResult.safe();

    const tables = this.extractTables(stmt);

    for (const table of tables) {
      const tableLower = table.toLowerCase();
      if (!this.allowedTables.has(tableLower)) {
        return ValidationResult.unsafe(
          `Table '${table}' is not in allowed tables list`
        );
      }
    }

    return ValidationResult.safe();
  }

  private extractTables(stmt: AST): string[] {
    const tables: string[] = [];
    this.walkAst(stmt, (node: any) => {
      // Handle table references
      if (node && typeof node === 'object') {
        // Direct table reference
        if (node.table && typeof node.table === 'string') {
          tables.push(node.table);
        }
        // FROM clause with array of tables
        if (Array.isArray(node.from)) {
          for (const item of node.from) {
            if (item?.table) tables.push(item.table);
          }
        }
      }
    });
    return [...new Set(tables)]; // Deduplicate
  }

  private extractAllTables(statements: AST[]): string[] {
    const allTables: string[] = [];
    for (const stmt of statements) {
      allTables.push(...this.extractTables(stmt));
    }
    return [...new Set(allTables)];
  }

  private walkAst(node: any, callback: (node: any) => void): void {
    if (!node || typeof node !== 'object') return;

    callback(node);

    if (Array.isArray(node)) {
      for (const item of node) {
        this.walkAst(item, callback);
      }
    } else {
      for (const key of Object.keys(node)) {
        this.walkAst(node[key], callback);
      }
    }
  }
}

File: packages/python/proxql/validator.py
"""Main Validator class - the primary entry point for ProxQL."""

from __future__ import annotations

from .cost import CostEstimate, CostEstimator, CostLevel
from .exceptions import ParseError
from .limits import LimitEnforcer
from .parser import Parser
from .policy import Mode, PolicyEngine
from .result import ValidationResult
from .security import SecurityChecker, SecurityConfig


class Validator:
    """SQL validator that checks queries against configurable policies.

    This is the main entry point for ProxQL. Create a Validator with your
    desired configuration, then call validate() on SQL strings.

    The validation process has multiple phases:
    1. Policy checks: Statement types and table allowlists (fast)
    2. Security rules: Deep AST analysis for injection patterns
    3. Limit enforcement: Row limit checks (optional)
    4. Cost estimation: Query complexity analysis (optional)

    Example:
        >>> validator = Validator(mode="read_only")
        >>> result = validator.validate("SELECT * FROM users")
        >>> result.is_safe
        True

        >>> result = validator.validate("DROP TABLE users")
        >>> result.is_safe
        False

        # With row limits
        >>> validator = Validator(mode="read_only", max_rows=1000)
        >>> result = validator.validate("SELECT * FROM users LIMIT 5000")
        >>> result.is_safe
        False

        # With cost estimation
        >>> validator = Validator(mode="read_only", estimate_cost=True)
        >>> result = validator.validate("SELECT * FROM a JOIN b JOIN c JOIN d")
        >>> result.cost.level
        <CostLevel.HIGH: 3>
    """

    def __init__(
        self,
        mode: Mode | str = Mode.READ_ONLY,
        allowed_tables: set[str] | list[str] | None = None,
        allowed_statements: set[str] | list[str] | None = None,
        blocked_statements: set[str] | list[str] | None = None,
        dialect: str | None = None,
        security_config: SecurityConfig | None | bool = True,
        max_rows: int | None = None,
        require_limit: bool = False,
        estimate_cost: bool = False,
        block_high_cost: bool = False,
        max_cost_level: CostLevel | str | None = None,
    ) -> None:
        """Initialize the validator.

        Args:
            mode: Validation mode. Options:
                - "read_only": Only SELECT allowed (default)
                - "write_safe": SELECT, INSERT, UPDATE allowed
                - "custom": Use allowed_statements/blocked_statements
            allowed_tables: Optional whitelist of accessible table names.
            allowed_statements: For custom mode, statements to allow.
            blocked_statements: For custom mode, statements to block.
            dialect: SQL dialect ('postgres', 'mysql', 'snowflake', etc).
            security_config: Security rule configuration. Options:
                - True (default): Default config (HIGH+ severity)
                - SecurityConfig: Custom configuration
                - None/False: Disable security rules
            max_rows: Maximum allowed LIMIT value. Queries with LIMIT > max_rows
                      are rejected. None means no limit.
            require_limit: If True, SELECT queries without LIMIT are rejected.
            estimate_cost: If True, estimate query cost/complexity.
            block_high_cost: If True, block queries with HIGH or EXTREME cost.
            max_cost_level: Maximum allowed cost level. Queries exceeding this
                           are rejected. Can be CostLevel or string like "MEDIUM".
        """
        self._parser = Parser(dialect=dialect)
        self._dialect = dialect
        self._policy = PolicyEngine(
            mode=mode,
            allowed_tables=allowed_tables,
            allowed_statements=allowed_statements,
            blocked_statements=blocked_statements,
        )

        # Set up security checker
        self._security: SecurityChecker | None
        if security_config is True:
            self._security = SecurityChecker(SecurityConfig())
        elif security_config is False or security_config is None:
            self._security = None
        else:
            self._security = SecurityChecker(security_config)

        # Set up limit enforcer
        self._limit_enforcer: LimitEnforcer | None = None
        if max_rows is not None or require_limit:
            self._limit_enforcer = LimitEnforcer(
                max_rows=max_rows,
                require_limit=require_limit,
            )

        # Set up cost estimator
        self._estimate_cost = estimate_cost or block_high_cost or max_cost_level is not None
        self._cost_estimator: CostEstimator | None = None
        if self._estimate_cost:
            self._cost_estimator = CostEstimator()

        # Cost blocking settings
        self._block_high_cost = block_high_cost
        self._max_cost_level: CostLevel | None = None
        if max_cost_level is not None:
            if isinstance(max_cost_level, str):
                self._max_cost_level = CostLevel[max_cost_level.upper()]
            else:
                self._max_cost_level = max_cost_level
        elif block_high_cost:
            self._max_cost_level = CostLevel.MEDIUM

    def validate(self, sql: str) -> ValidationResult:
        """Validate a SQL query string.

        Parses the SQL and checks it against:
        1. Configured policies (statement types, table allowlists)
        2. Security rules (injection patterns, dangerous functions, etc.)
        3. Row limit enforcement (if configured)
        4. Cost estimation (if configured)

        Multi-statement queries are all validated; if any statement fails,
        the entire query is considered unsafe.

        Args:
            sql: The SQL query string to validate.

        Returns:
            ValidationResult with is_safe=True if the query passes all checks,
            or is_safe=False with a reason if blocked.

        Note:
            Parse errors are treated as unsafe (is_safe=False) with the
            parse error message as the reason.
        """
        # Handle empty/whitespace input
        sql = sql.strip()
        if not sql:
            return ValidationResult(
                is_safe=False,
                reason="Empty query",
            )

        # Try to parse
        try:
            statements = self._parser.parse(sql)
        except ParseError as e:
            return ValidationResult(
                is_safe=False,
                reason=f"Parse error: {e}",
            )

        if not statements:
            return ValidationResult(
                is_safe=False,
                reason="No valid SQL statements found",
            )

        # Validate each statement
        all_tables: list[str] = []
        all_warnings: list[str] = []
        combined_cost: CostEstimate | None = None
        limit_value: int | None = None

        for stmt in statements:
            statement_type = self._parser.get_statement_type(stmt)
            tables = self._parser.extract_tables(stmt)
            all_tables.extend(tables)

            # Phase 1: Policy checks (statement type, table allowlist)
            result = self._policy.evaluate(stmt, statement_type, tables)
            if not result.is_safe:
                return result

            # Phase 2: Security rule checks
            if self._security is not None:
                sec_result = self._security.check(stmt, dialect=self._dialect, raw_sql=sql)
                if not sec_result.passed:
                    violation = sec_result.first_violation
                    if violation is not None and violation.message:
                        msg = violation.message
                    else:
                        msg = "Security check failed"
                    return ValidationResult(
                        is_safe=False,
                        reason=msg,
                        statement_type=statement_type,
                        tables=tables,
                    )

            # Phase 3: Limit enforcement
            if self._limit_enforcer is not None:
                limit_result = self._limit_enforcer.check(stmt)
                if not limit_result.is_ok:
                    return ValidationResult(
                        is_safe=False,
                        reason=limit_result.reason,
                        statement_type=statement_type,
                        tables=tables,
                    )
                if limit_result.limit_value is not None:
                    limit_value = limit_result.limit_value

            # Phase 4: Cost estimation
            if self._cost_estimator is not None:
                cost = self._cost_estimator.estimate(stmt)

                # Track highest cost across statements
                if combined_cost is None or cost.level > combined_cost.level:
                    combined_cost = cost

                # Check if cost exceeds maximum
                if self._max_cost_level is not None and cost.level > self._max_cost_level:
                    return ValidationResult(
                        is_safe=False,
                        reason=f"Query cost ({cost.level.name}) exceeds maximum allowed ({self._max_cost_level.name})",
                        statement_type=statement_type,
                        tables=tables,
                        cost=cost,
                    )

                # Add warnings for high-cost queries that aren't blocked
                if cost.level >= CostLevel.HIGH and self._max_cost_level is None:
                    all_warnings.append(
                        f"High cost query ({cost.level.name}): {', '.join(cost.factors)}"
                    )

        # All statements passed
        # Return result with combined info from all statements
        final_type = (
            self._parser.get_statement_type(statements[0]) if len(statements) == 1 else "MULTI"
        )
        return ValidationResult(
            is_safe=True,
            statement_type=final_type,
            tables=list(set(all_tables)),
            cost=combined_cost,
            limit_value=limit_value,
            warnings=all_warnings,
        )
